File: 1358-number-of-substrings-containing-all-three-characters/solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        # Dictionary to store the frequency of 'a', 'b', and 'c'
        count = {'a': 0, 'b': 0, 'c': 0}
        left = 0
        result = 0
        n = len(s)
        
        for right in range(n):
            # Expand the window by adding the character at 'right'
            count[s[right]] += 1
            
            # While the window contains all three characters
            while count['a'] > 0 and count['b'] > 0 and count['c'] > 0:
                # If s[left...right] is valid, then s[left...right], 
                # s[left...right+1], ..., s[left...n-1] are valid substrings
                # ending at or after 'right'.
                # Actually, the logic is: for a fixed 'right', if the window
                # [left, right] is valid, all substrings starting at 0..left
                # ending at 'right' are valid.
                result += (n - right)
                
                # Shrink the window from the left to find the next valid window
                count[s[left]] -= 1
                left += 1
                
        return result

File: 1358-number-of-substrings-containing-all-three-characters/solution.cpp
#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    int numberOfSubstrings(std::string s) {
        // frequency array to store count of 'a', 'b', and 'c'
        int count[3] = {0, 0, 0};
        int left = 0, result = 0;
        int n = s.length();

        for (int right = 0; right < n; ++right) {
            // Update the count for the character at the right boundary
            count[s[right] - 'a']++;

            // While the current window is valid (has at least one of each)
            while (count[0] > 0 && count[1] > 0 && count[2] > 0) {
                // If s[left...right] is valid, then all s[i...right] where 
                // i <= left are also valid. There are (left + 1) such substrings.
                result += (n - right);
                
                // Shrink the window from the left
                count[s[left] - 'a']--;
                left++;
            }
        }
        return result;
    }
};